src/lib/Interpreter.hs

```diff
@@ -94,60 +106,74 @@ evalOpDefined expr = case expr of
     _ -> error $ "FFI function not recognized: " ++ name
-  ArrayOffset arrArg _ offArg -> Con $ ArrayLit (ArrayTy b) (arrayOffset arr $ fromIntegral off)
-    where (ArrayVal (ArrayTy (TabTy _ b)) arr, IdxRepVal off) = (arrArg, offArg)
-  ArrayLoad arrArg -> Con $ Lit $ arrayHead arr where (ArrayVal (ArrayTy (BaseTy _)) arr) = arrArg
+  PtrOffset (Con (Lit (PtrLit (_, a, t) p))) (IdxRepVal i) ->
+    return $ Con $ Lit $ PtrLit (DerivedPtr, a, t) $ p `plusPtr` (sizeOf t * fromIntegral i)
+  PtrLoad (Con (Lit (PtrLit (_, Heap CPU, t) p))) -> (Con . Lit) <$> loadLitVal p t
+  PtrLoad (Con (Lit (PtrLit (_, a, _) _))) ->
+    error $ "Can't load from this address space in interpreter: " ++ pprint a
   IndexAsInt idxArg -> case idxArg of
-    Con (Coerce (TC (IntRange   _ _  )) i) -> i
-    Con (Coerce (TC (IndexRange _ _ _)) i) -> i
-    Con (AnyValue t)                       -> anyValue t
-    _ -> evalEmbed (indexToIntE (getType idxArg) idxArg)
-  Fst p         -> x                     where (PairVal x _) = p
-  Snd p         -> y                     where (PairVal _ y) = p
+    Con (IntRangeVal   _ _   i) -> return i
+    Con (IndexRangeVal _ _ _ i) -> return i
+    Con (AnyValue t)                       -> return $ anyValue t
+    _ -> evalEmbed (indexToIntE idxArg)
+  Fst p -> return x  where (PairVal x _) = p
+  Snd p -> return y  where (PairVal _ y) = p
   _ -> error $ "Not implemented: " ++ pprint expr
 
-indices :: Type -> [Atom]
-indices ty = case ty of
-  TC (IntRange _ _)      -> fmap (Con . Coerce ty . IdxRepVal) [0..(fromIntegral $ n - 1)]
-  TC (IndexRange _ _ _)  -> fmap (Con . Coerce ty . IdxRepVal) [0..(fromIntegral $ n - 1)]
-  TC (PairType lt rt)    -> [PairVal l r | l <- indices lt, r <- indices rt]
-  TC (UnitType)          -> [UnitVal]
-  RecordTy (NoExt types) -> let
-    subindices = map indices (toList types)
-    -- Earlier indices change faster than later ones, so we need to first
-    -- iterate over the current index and then over all previous ones. For
-    -- efficiency we build the indices in reverse order and then reassign them
-    -- at the end with `reverse`.
-    addAxisInReverse prevs curs = [cur:prev | cur <- curs, prev <- prevs]
-    products = foldl addAxisInReverse [[]] subindices
-    in map (\idxs -> Record $ restructure (reverse idxs) types) products
-  VariantTy (NoExt types) -> let
-    subindices = fmap indices types
-    reflect = reflectLabels types
-    zipped = zip (toList reflect) (toList subindices)
-    in concatMap (\((label, i), args) ->
-      Variant (NoExt types) label i <$> args) zipped
-  _ -> error $ "Not implemented: " ++ pprint ty
-  where n = indexSetSize ty
+-- We can use this when we know we won't be dereferencing pointers. A better
+-- approach might be to have a typeclass for the pointer dereferencing that the
+-- interpreter does, with a dummy instance that throws an error if you try.
+indicesNoIO :: Type -> [Atom]
+indicesNoIO = unsafePerformIO . indices
```

Comment:
When would creating the list of indices ever dereference a pointer? I'm pretty sure there's no wait for pointers to appear in the type?

Reply:
Hopefully never!